packages/story-editor/src/app/pageCanvas/pageCanvasProvider.js

```diff
@@ -0,0 +1,205 @@
+   *
+   * @return {void}
+   */
+  const generateDeferredCurrentPageCanvas = useCallback(() => {
+    const { currentPageValue, pageCanvasMapValue } = valuesRef.current;
+    const canvas = pageCanvasMapValue[currentPageValue.id];
+    if (!canvas) {
+      generateDeferredPageCanvas([currentPageValue.id, currentPageValue]);
+    }
+  }, [generateDeferredPageCanvas]);
+
+  /**
+   * Gets or creates a canvas from the page and excludes
+   * the current selection from the generated canvas.
+   *
+   * @return {HTMLCanvasElement} generated canvas
+   */
+  const getSelectionExclusionCanvas = useCallback(
+    async (page, selection) => {
+      const pageWithoutSelection = getPageWithoutSelection(page, selection);
+      let canvas = getSnapshotCanvas(pageWithoutSelection);
+      if (!canvas) {
+        // Generate the page canvas with the excluded elements if we don't already
+        // have a valid canvas in the cache.
+        canvas = await storyPageToCanvas(pageWithoutSelection, {});
+        setSnapshot({
+          page: pageWithoutSelection,
+          canvas,
+        });
+      }
+      return canvas;
+    },
+    [setSnapshot, getSnapshotCanvas]
+  );
+
+  /**
+   * Gets or creates a generated canvas from the current story page.
+   *
+   * @return {HTMLCanvasElement} generated canvas
+   */
+  const getCanvas = useCallback(
+    async (page) => {
+      const { pageCanvasMapValue } = valuesRef.current;
+
+      let canvas = pageCanvasMapValue[page.id];
+      if (!canvas) {
+        canvas = await storyPageToCanvas(page, {});
+        actions.setPageCanvas({ pageId: page.id, canvas });
+      }
+      return canvas;
+    },
+    [actions]
+  );
+
+  /**
+   * Given an element, returns accessible text colors relative
+   * to the current page
+   *
+   * @param {} story element
+   * @return {Object} Returns object consisting of color and backgroundColor in case relevant.
+   */
+  const calculateAccessibleTextColors = useCallback(
+    async (element) => {
+      const { currentPageValue, singleElementSelectionValue } =
+        valuesRef.current;
+
+      let canvas;
+      if (singleElementSelectionValue.includes(element.id)) {
+        canvas = await getSelectionExclusionCanvas(
+          currentPageValue,
+          singleElementSelectionValue
+        );
+      } else {
+        canvas = await getCanvas(currentPageValue);
+      }
+
+      const { fontSize } = element;
+      const pixelData = getPixelDataFromCanvas(canvas, element);
+      const accessibleTextColors = getAccessibleTextColorsFromPixels(
+        pixelData,
+        fontSize
+      );
+
+      return accessibleTextColors;
+    },
+    [getSelectionExclusionCanvas, getCanvas]
+  );
+
+  return (
+    <Context.Provider
+      value={{
+        state: {
+          pageCanvasMap,
+        },
+        actions: {
+          calculateAccessibleTextColors,
+          generateDeferredPageCanvas,
+          generateDeferredCurrentPageCanvas,
+        },
+      }}
```

Comment:
💭 does it matter if the object on `178` is being re-created every render? I guess it shouldn't matter as long as the references in `state` and `actions` are being pulled out of the object, and the parent object reference isn't being compared directly in children, right?

Reply:
correct. If we're using selectors well, there's no difference if you were to memoize the object held in the providers value